File: supercoolQL/src/main/com/matthewchapman/ql/ast/statement/IfStatement.java
package com.matthewchapman.ql.ast.statement;

import com.matthewchapman.ql.ast.Expression;
import com.matthewchapman.ql.ast.Statement;
import com.matthewchapman.ql.validator.QLVisitor;

import java.util.ArrayList;

/**
 * Created by matt on 21/02/2017.
 */

public class IfStatement extends Statement {

    private final ArrayList<Statement> statements;
    private Expression expression;

    public IfStatement(Expression e, ArrayList<Statement> s) {
        this.statements = new ArrayList<>();
        this.addStatements(s);
        this.expression = e;
    }

    public void addStatements(ArrayList<Statement> statements) {
        for (Statement s : statements) {
            this.statements.add(s);
        }
    }

    public void setExpression(Expression e) {
        this.expression = e;
    }

    public ArrayList<Statement> getStatements() {
        return this.statements;
    }

    public Expression getExpression() {
        return this.expression;
    }

    @Override
    public <T> T accept(QLVisitor<T> visitor) {
        return visitor.visit(this);
    }

}

File: OffByOne/src/OffByOne.Ql/Ast/Statements/IfStatement.cs
namespace OffByOne.Ql.Ast.Statements
{
    using System;
    using System.Collections.Generic;

    using OffByOne.Ql.Ast.Expressions;
    using OffByOne.Ql.Ast.Statements.Base;
    using OffByOne.Ql.Common.Visitors.Contracts;

    public class IfStatement : Statement
    {
        public IfStatement(
            Expression condition,
            IEnumerable<Statement> statements,
            IEnumerable<Statement> elseStatements)
        {
            if (condition == null)
            {
                throw new ArgumentNullException(nameof(condition));
            }

            if (statements == null)
            {
                throw new ArgumentNullException(nameof(condition));
            }

            this.Condition = condition;
            this.Statements = statements;
            this.ElseStatements = elseStatements;
        }

        public Expression Condition { get; }

        public IEnumerable<Statement> Statements { get; }

        public IEnumerable<Statement> ElseStatements { get; }

        public override TResult Accept<TResult, TContext>(
            IStatementVisitor<TResult, TContext> visitor,
            TContext environment)
        {
            return visitor.Visit(this, environment);
        }

        public override ISet<string> GetDependencies()
        {
            return this.Condition.GetDependencies();
        }

        public bool IsComputable(string key)
        {
            return this.GetDependencies().Contains(key);
        }
    }
}
